File: src/casper_ffi.mjs
import { BitArray } from "./gleam.mjs";
import { Ok, Error } from "./gleam.mjs";
import * as crypto from "node:crypto";

export function strong_random_bytes(n) {
  return new BitArray(crypto.randomBytes(32));
}

export function encrypt_internal(input, key) {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('chacha20-poly1305', key.rawBuffer, iv);
  let encrypted = cipher.update(input.rawBuffer);
  cipher.final();
  const tag = cipher.getAuthTag();
  return new BitArray(Buffer.concat([iv, tag, Buffer.from(encrypted)]));
}

export function encrypt_with_internal(input, associatedData, key) {
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv('chacha20-poly1305', key.rawBuffer, iv);
  cipher.setAAD(associatedData.rawBuffer);
  let encrypted = cipher.update(input.rawBuffer);
  cipher.final();
  const tag = cipher.getAuthTag();
  return new BitArray(Buffer.concat([iv, tag, Buffer.from(encrypted)]));
}

export function decrypt_internal(input, key) {
  try {
    const iv = Buffer.copyBytesFrom(input.rawBuffer, 0, 12);
    const tag = Buffer.copyBytesFrom(input.rawBuffer, 12, 16);
    const encrypted = Buffer.copyBytesFrom(input.rawBuffer, 28);
    const decipher = crypto.createDecipheriv('chacha20-poly1305', key.rawBuffer, iv);
    decipher.setAuthTag(tag);
    let decrypted = decipher.update(encrypted);
    decipher.final();
    return new Ok(new BitArray(decrypted));
  } catch {
    return new Error(null);
  }
}

export function decrypt_with_internal(input, associatedData, key) {
  try {
    const iv = Buffer.copyBytesFrom(input.rawBuffer, 0, 12);
    const tag = Buffer.copyBytesFrom(input.rawBuffer, 12, 16);
    const encrypted = Buffer.copyBytesFrom(input.rawBuffer, 28);
    const decipher = crypto.createDecipheriv('chacha20-poly1305', key.rawBuffer, iv);
    decipher.setAuthTag(tag);
    decipher.setAAD(associatedData.rawBuffer);
    let decrypted = decipher.update(encrypted);
    decipher.final();
    return new Ok(new BitArray(decrypted));
  } catch {
    return new Error(null);
  }
}

File: src/casper_ffi.erl
-module(casper_ffi).

-export([
    generate_key/0,
    encrypt_internal/2,
    decrypt_internal/2,
    encrypt_with_internal/3,
    decrypt_with_internal/3
]).

generate_key() ->
    crypto:strong_rand_bytes(32).

encrypt_internal(Value, Key) ->
    Iv = crypto:strong_rand_bytes(12),
    {Encrypted, Tag} = crypto:crypto_one_time_aead(
        chacha20_poly1305, Key, Iv, Value, <<>>, true
    ),
    <<Iv/binary, Tag/binary, Encrypted/binary>>.

decrypt_internal(<<Iv:12/binary, Tag:16/binary, Encrypted/binary>>, Key) ->
    Decrypted = crypto:crypto_one_time_aead(
        chacha20_poly1305, Key, Iv, Encrypted, <<>>, Tag, false
    ),
    case Decrypted of
        {_, _} -> {error, nil};
        error -> {error, nil};
        {error, _, _} -> {error, nil};
        Decrypt -> {ok, Decrypt}
    end.

encrypt_with_internal(Value, AssociatedData, Key) ->
    Iv = crypto:strong_rand_bytes(12),
    {Encrypted, Tag} = crypto:crypto_one_time_aead(
        chacha20_poly1305, Key, Iv, Value, AssociatedData, true
    ),
    <<Iv/binary, Tag/binary, Encrypted/binary>>.

decrypt_with_internal(<<Iv:12/binary, Tag:16/binary, Encrypted/binary>>, AssociatedData, Key) ->
    Decrypted = crypto:crypto_one_time_aead(
        chacha20_poly1305, Key, Iv, Encrypted, AssociatedData, Tag, false
    ),
    case Decrypted of
        {_, _} -> {error, nil};
        error -> {error, nil};
        {error, _, _} -> {error, nil};
        Decrypt -> {ok, Decrypt}
    end.
